db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go

```diff
@@ -0,0 +1,255 @@
+
+// RemoveDuplicatedTenantIdsOrgIds fetches all the duplicated tenants, makes sure that they are indeed duplicated, and
+// unifies all the subresources in a single tenant before deleting the duplicated one.
+func RemoveDuplicatedTenantIdsOrgIds() *gormigrate.Migration {
+	return &gormigrate.Migration{
+		ID: "20220608090500",
+		Migrate: func(db *gorm.DB) error {
+			logging.Log.Info(`Migration "remove duplicated tenants" started`)
+			defer logging.Log.Info(`Migration "remove duplicated tenants" ended`)
+
+			// Updatable tables contains all the table names which have the "tenant_id" column which must be updated.
+			updatableTables := []string{
+				"application_authentications",
+				"applications",
+				"authentications",
+				"endpoints",
+				"sources",
+				"source_rhc_connections",
+				"users",
+			}
+
+			// This query selects all the tenants which have duplicated "external_tenant" or "org_id" column values.
+			duplicatesSql := `
+				SELECT
+					"t"."id", "t"."name", "t"."description", "t"."external_tenant", "t"."org_id", "t"."created_at", "t"."updated_at"
+				FROM
+					"tenants" AS "t"
+				WHERE
+					"t"."external_tenant" IN (
+						SELECT
+							"texten"."external_tenant"
+						FROM
+							"tenants" AS "texten"
+						GROUP BY
+							"texten"."external_tenant"
+						HAVING
+							count("texten"."external_tenant") > 1
+						AND
+							"texten"."external_tenant" != ''
+					)
+				OR
+					"t"."org_id" IN (
+						SELECT
+							"torgid"."org_id"
+						FROM
+							"tenants" AS "torgid"
+						GROUP BY
+							"torgid"."org_id"
+						HAVING
+							count("torgid"."org_id") > 1
+						AND
+							"torgid"."org_id" != ''
+					)
+			`
+
+			// The idea is that the structure of the following maps will be as follows:
+			//
+			// externalTenantIds:
+			// 	"12345" -> [12, 5123, 123]
+			// 	"4717"	-> [47, 511, 511]
+			//
+			// orgIdIds:
+			// 	"abcde-12345"	-> [51, 98]
+			// 	"dfhea-4918"	-> [5811, 368128]
+			var externalTenantIds = make(map[string][]int64)
+			var orgIdIds = make(map[string][]int64)
+
+			// The migration updates the subresources of a tenant. It first groups all the tenant IDs with their
+			// duplicated "external_tenant" and "org_id". Once it has those, it loops through every table that depends
+			// on the "tenants" table, and updates the "tenant_id" column of the resources from the duplicated tenants
+			// to the main tenant ID. Finally, it updates the "tenants" table by marking the duplicated tenants with
+			// an easily identifiable "external_tenant" and "org_id" values.
+			err := db.Transaction(func(tx *gorm.DB) error {
+				var tenants []model.Tenant
+
+				err := tx.Debug().Raw(duplicatesSql).Scan(&tenants).Error
+				if err != nil {
+					return err
+				}
+
+				// Group all "external_tenant" and "org_id"s into the above defined maps.
+				for _, t := range tenants {
+					// Append the ID to the external tenants map.
```

Comment:
maybe log message here ? to see which tenants are duplicate

Reply:
Thank you :)